arch/arm/core/cortex_m/prep_c.c

```diff
@@ -119,31 +119,47 @@ static inline void enable_floating_point(void)
 {
 	/*
 	 * Upon reset, the Co-Processor Access Control Register is 0x00000000.
-	 * Enable CP10 and CP11 co-processors to enable floating point.
+	 * Enable CP10 and CP11 Co-Processors to enable access to floating
+	 * point registers.
 	 */
+#if defined(CONFIG_USERSPACE)
+	/* Full access */
 	SCB->CPACR |= CPACR_CP10_FULL_ACCESS | CPACR_CP11_FULL_ACCESS;
+#else
+	/* Privileged access only */
+	SCB->CPACR |= CPACR_CP10_PRIV_ACCESS | CPACR_CP11_PRIV_ACCESS;
+#endif /* CONFIG_USERSPACE */
 	/*
 	 * Upon reset, the FPU Context Control Register is 0xC0000000
 	 * (both Automatic and Lazy state preservation is enabled).
+	 */
+#if !defined(CONFIG_FP_SHARING)
+	/* Default mode is Unshared FP registers mode. We disable the
+	 * the automatic stacking of FP registers (automatic setting
+	 * of FPCA bit in CONTROL register), upon exception entries,
+	 * as the FP registers are to be used by a single context (and
+	 * the use of FP registers in ISRs is not supported). This
+	 * configuration improves interrupt latency and decreases the
+	 * stack memory requirement for the thread that makes use of
+	 * the FP co-processor.
+	 */
+	FPU->FPCCR &= (~(FPU_FPCCR_ASPEN_Msk | FPU_FPCCR_LSPEN_Msk));
```

Comment:
I would not rely on the initial contents of the FPCCR.

Reply:
> 
> 
> I would add one thing: In the unshared mode only one entity should use the FPU. But since we are not checking that, the FPU state could be easily corrupted leading to errors which are hard to trace.
> Could we have an mechanism which tell the user (ASSERT) that FPU is used by more than one entity if unshared mode is used?

This is interesting feedback.

So, the whole point of this re-work of the Un-shared FP Services relies on the assumption that the user knows what they are doing. This allowed us to do all these optimizations (memory usage, interrupt latency, simplified context-switch, etc.). If we do not trust the user, the whole idea of un-shared FP context is invalid. :)

Also, in order to add the proposed ASSERT mechanism, we need to know, somehow, that a thread entity is using the FPU. This is problematic, because we have disabled the automatic-state-preservation [`ASPEN`] (to improve latency). This means that the CONTROL.FPCA will never be set to 1.

Therefore, we would need to save the state manually (only when we compile with ASSERTS ON) for example in the `thread.arch` container. We would need an API for that, which we do not have, that is the first problem.

But say, we could add a bit flag in the `thread.arch.mode`, for instance - I am not against that, and in fact, this might be great for the **Sharing FP Services** use-case!

But for now, this would mean that thread.arch struct would look differently when ASSERTS are on; we do not do these things, usually, i the core code.

Anyways, if we would agree to do this, then, yes, we could, theoretically enable the `ASPEN` inside context-switching, for threads with "FP-mode" set, and check the CONTROL.FCPA value when we switch-back to this thread; it it would be 1, then the FPU state is corrupted and we would ASSERT.

Does it make sense to do this, though? We assume user selected Un-shared FP Services and know what he is doing.

>But since we are not checking that, the FPU state could be easily corrupted leading to errors which are hard to trace.

@pizi-nordic, idea: assume a user sets `CONFIG_FLOAT=y` (un-shared), but sees irregularities with the FPU, so they suspect it gets corrupted by other context. All they have to do is to enable `CONFIG_FP_SHARING`; which allows the application to safely share the FPU among multiple threads. If the observed irregularities go away, then the user knows that the FPU was indeed corrupted by other context.

Reply:
Regarding FPCCR: The only RW bits are the ASPEN, LSPEN, so I think the operation is safe.

Reply:
The problem is that these irregularities could be hard to spot. For example they could manifest itself as some distortions/invalid frame data in audio compressed by the codec using FPU which happens only when something is specific executed during compression. In such case checking the CONFIG_FP_SHARING will be the last thing the user will think about.

The benefit of having non-shared FPU is obvious, but we should keep Zephyr usage simple.
My proposal is:
- Keep CONFIG_FP_SHARING=y as default, with big, fat warning that you have to understand all implications before considering changing that.
- Add CONFIG_FP_UNSHARING_TEST which allow you to test if you app is compatible with disabled CONFIG_FP_SHARING...

Reply:
>Keep CONFIG_FP_SHARING=y as default, with big, fat warning that you have to understand all implications before considering changing that.

The default, now, is Un-shared FP. For Shared, we need to set both CONFIG_FLOAT and CONFIG_FP_SHARING. I am not against changing the default to Shared, but this is outside the scope of this PR. 

>Add CONFIG_FP_UNSHARING_TEST which allow you to test if you app is compatible with disabled CONFIG_FP_SHARING...

So what it the value of this? What you propose looks like:
1. The user sees issues with their code.
2. The user enables ASSERTS for debugging.
3. The user suspects problems with multi-usage of FPU.
4. The user enables CONFIG_FP_UNSHARING_TEST to see if there are multiple entities using the FPU. The option will trigger ASSERTs on corruptions due to multi-usage. **Complex** to implement.

I propose something simpler:
1. The user sees issues with their code.
2. The user enables ASSERTS for debugging.
3. The user suspects problems with multi-usage of FPU.
4. The user enables CONFIG_FP_SHARING to see if there are multiple entities using the FPU.
   The option will make any "issues" related to multi-usage corruptions **disappear**. 
   **No** implementation.
   Issues disappear? It was a multi-usage corruption. Issues still present? It was not a multi-usage corruption.

Reply:
I would prefer opposite way of thinking. Instead of waiting for issues and then actively searching the cause, we should prevent them: So for me, the right order will be developing the app with all debugging options enabled, and then if everything works disabling the checks. Such approach saves a lot of time.

For me the biggest problem with lack of checking in non-shared FPU is that the errors will manifest itself in a way which will not suggest FPU issue. For example you might get invalid data from sensor from time to time and I can guarantee you that you will start from tracing communications with the sensor as well as checking its performance.

Reply:
Sure, let's formulate this, then, in the opposite way:
1. The user implements an application that uses the FPU. Sets FP_SHARING, ASSERTS, etc. during development stage.
2. The user sees no issues and finalizes development.
3. The user removes FP_SHARING, as they are confident that the FPU is used in single context.
4. The user spots "issues" related or un-related to the FPU.
5. The user goes back to development and sets FP_SHARING. Issues disappear. This is a controlled experiment.
6. The user concludes that the FPU context is corrupted.

I see the value of testing corruption in un-shared mode, (exactly as I see the value of any other functional testing), but we must also consider code-readability and implementation complexity. 

In any case, what you propose can be implemented right on top of this PR. I will file an issue for 1.15. I am really not against the proposal - it won't break anything AFAICS, but it will take time to implement and is of little value, IMHO.

Reply:
IHMO this is the right order, but instead of "issues" in point 4, we should see asserts (the part which verified the setup might be later disabled if no issues were reported).

Reply:
Yeap. Seeing ASSERTS instead of "issues-to-be-resolved-by-FP-Sharing" seems to be the only value of the proposal, IMHO. I can buy it. See, now #15292.

Reply:
> Seeing ASSERTS instead of "issues-to-be-resolved-by-FP-Sharing" seems to be the only value of the proposal, IMHO.

And this is what I proposed :)

Reply:
Resolving this, since an issue for adding ASSERT-based testing of Unshared-FP-mode is created.
@pizi-nordic , pls, re-review.